File: internal/meta/meta.go
// Package meta provides typed accessors for the reserved sections of
// the per-task `metadata` JSON blob (tasks.metadata).
//
// Most of metadata is free-form, but the engine reserves a small set of
// top-level keys for its own use. Today there is exactly one: the
// `step_visits` map, which tracks how many times a task has entered
// each workflow step (see docs/plans/20260520-Step-Visit-Limits.md).
//
// Storing reserved data through these helpers (instead of indexing the
// raw map by string) keeps the engine from accidentally corrupting the
// shape after a json.Unmarshal round-trip widens int → float64.
package meta

import "fmt"

// StepVisitsKey is the reserved top-level metadata key whose value is a
// map step_id → visit count (>=0). Other tools should treat anything
// under this key as engine-owned.
const StepVisitsKey = "step_visits"

// StepVisits is the typed shape of metadata[StepVisitsKey]. Keys are
// step ids ("st-XXXX"); values are visit counts.
type StepVisits map[string]int

// GetStepVisits returns a typed copy of the reserved step_visits map.
// Tolerates the wire shape produced by json.Unmarshal — integer values
// are decoded as float64 by default. Unknown / malformed values are
// dropped silently to keep the engine resilient to free-form edits.
//
// Returns a non-nil, possibly empty map.
func GetStepVisits(m map[string]any) StepVisits {
	out := make(StepVisits)
	if m == nil {
		return out
	}
	raw, ok := m[StepVisitsKey]
	if !ok || raw == nil {
		return out
	}
	rawMap, ok := raw.(map[string]any)
	if !ok {
		// Some other tool wrote a non-object under step_visits. Treat as
		// "no recorded visits"; the engine never reads it that way and
		// MutateStepVisits below will overwrite it on the next bump.
		return out
	}
	for k, v := range rawMap {
		switch n := v.(type) {
		case float64:
			out[k] = int(n)
		case int:
			out[k] = n
		case int64:
			out[k] = int(n)
		}
	}
	return out
}

// SetStepVisits replaces the reserved step_visits sub-object on m with
// sv. The map is stored as `map[string]any` so json.Marshal produces
// the canonical wire shape; integer values round-trip correctly.
//
// Mutates m in place. Panics if m is nil. The store.Store contract
// promises every UpdateMetadata closure receives a non-nil map (see
// store.Store.UpdateMetadata docs), so the typical engine pattern
// "load, MutateStepVisits, save" cannot trip this.
func SetStepVisits(m map[string]any, sv StepVisits) {
	if m == nil {
		panic("meta.SetStepVisits: nil map")
	}
	if len(sv) == 0 {
		delete(m, StepVisitsKey)
		return
	}
	out := make(map[string]any, len(sv))
	for k, v := range sv {
		out[k] = v
	}
	m[StepVisitsKey] = out
}

// MutateStepVisits is the read-modify-write convenience used by the
// engine and the CLI. It hands fn a typed StepVisits, then writes the
// result back. An empty resulting map deletes the reserved key entirely
// so an empty metadata blob round-trips back to NULL.
//
// Panics if m is nil. The store.Store.UpdateMetadata contract
// guarantees the closure receives a non-nil map; this is what makes
// the panic unreachable from real engine code.
func MutateStepVisits(m map[string]any, fn func(StepVisits)) {
	if m == nil {
		panic("meta.MutateStepVisits: nil map")
	}
	sv := GetStepVisits(m)
	fn(sv)
	SetStepVisits(m, sv)
}

// ValidateStepVisitsLeaf returns nil iff v is a non-negative integer
// (or its float64 wire shape). Used by `autosk metadata set` to guard
// humans from corrupting the engine's view by writing a non-int leaf
// under step_visits.
func ValidateStepVisitsLeaf(v any) error {
	switch n := v.(type) {
	case int:
		if n < 0 {
			return fmt.Errorf("step_visits leaves must be >= 0, got %d", n)
		}
		return nil
	case int64:
		if n < 0 {
			return fmt.Errorf("step_visits leaves must be >= 0, got %d", n)
		}
		return nil
	case float64:
		if n < 0 || n != float64(int(n)) {
			return fmt.Errorf("step_visits leaves must be non-negative integers, got %v", v)
		}
		return nil
	default:
		return fmt.Errorf("step_visits leaves must be integers, got %T", v)
	}
}

// ValidateStepVisitsObject returns nil iff v is a JSON object whose
// every leaf passes ValidateStepVisitsLeaf. The CLI uses this when a
// human writes `metadata set --key step_visits --value '{...}'`.
func ValidateStepVisitsObject(v any) error {
	if v == nil {
		return nil
	}
	m, ok := v.(map[string]any)
	if !ok {
		return fmt.Errorf("step_visits must be a JSON object, got %T", v)
	}
	for k, leaf := range m {
		if err := ValidateStepVisitsLeaf(leaf); err != nil {
			return fmt.Errorf("step_visits[%q]: %w", k, err)
		}
	}
	return nil
}

File: crates/autosk-core/src/meta.rs
//! Typed accessors for the reserved `step_visits` section of a task's
//! free-form `metadata` JSON blob — the Rust port of `internal/meta`.
//!
//! The engine reserves the top-level key `step_visits` (a map `step_id ->
//! count`). Going through these helpers keeps a JSON round-trip from
//! corrupting the shape (Go widens int→float64; serde_json stores numbers
//! as [`serde_json::Number`], so the same defensive read applies).

use serde_json::{Map, Value};

/// Reserved top-level metadata key whose value is a map `step_id -> visit
/// count` (>= 0). Mirrors `meta.StepVisitsKey`.
pub const STEP_VISITS_KEY: &str = "step_visits";

/// Returns a typed copy of the reserved `step_visits` map. Tolerates the
/// wire shape (integers may decode as floats); malformed leaves are
/// dropped, matching `meta.GetStepVisits`. Always returns a (possibly
/// empty) map.
pub fn get_step_visits(m: &Map<String, Value>) -> std::collections::BTreeMap<String, i64> {
    let mut out = std::collections::BTreeMap::new();
    let Some(Value::Object(raw)) = m.get(STEP_VISITS_KEY) else {
        return out;
    };
    for (k, v) in raw {
        if let Some(n) = as_count(v) {
            out.insert(k.clone(), n);
        }
    }
    out
}

/// Replaces the reserved `step_visits` sub-object on `m` with `sv`. An
/// empty `sv` deletes the key entirely so an empty metadata blob
/// round-trips back to SQL NULL. Mirrors `meta.SetStepVisits`.
pub fn set_step_visits(m: &mut Map<String, Value>, sv: &std::collections::BTreeMap<String, i64>) {
    if sv.is_empty() {
        m.remove(STEP_VISITS_KEY);
        return;
    }
    let mut obj = Map::new();
    for (k, v) in sv {
        obj.insert(k.clone(), Value::from(*v));
    }
    m.insert(STEP_VISITS_KEY.to_string(), Value::Object(obj));
}

/// Read-modify-write helper used by the engine: hands `f` the typed
/// counters, then writes the result back (deleting the key when empty).
/// Mirrors `meta.MutateStepVisits`.
pub fn mutate_step_visits<F>(m: &mut Map<String, Value>, f: F)
where
    F: FnOnce(&mut std::collections::BTreeMap<String, i64>),
{
    let mut sv = get_step_visits(m);
    f(&mut sv);
    set_step_visits(m, &sv);
}

/// Interprets a JSON value as a non-negative visit count (whole numbers
/// only), or `None` when it is not a usable counter leaf.
fn as_count(v: &Value) -> Option<i64> {
    match v {
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                Some(i)
            } else {
                n.as_f64().and_then(|f| {
                    if f.fract() == 0.0 {
                        Some(f as i64)
                    } else {
                        None
                    }
                })
            }
        }
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn round_trip_and_mutate() {
        let mut m: Map<String, Value> = json!({"step_visits": {"st-a": 3, "st-b": 1}})
            .as_object()
            .unwrap()
            .clone();
        let sv = get_step_visits(&m);
        assert_eq!(sv["st-a"], 3);
        assert_eq!(sv["st-b"], 1);

        mutate_step_visits(&mut m, |sv| {
            *sv.entry("st-a".to_string()).or_insert(0) += 1;
        });
        assert_eq!(get_step_visits(&m)["st-a"], 4);
    }

    #[test]
    fn empty_deletes_key() {
        let mut m: Map<String, Value> = json!({"step_visits": {"st-x": 1}})
            .as_object()
            .unwrap()
            .clone();
        mutate_step_visits(&mut m, |sv| {
            sv.remove("st-x");
        });
        assert!(!m.contains_key(STEP_VISITS_KEY));
    }

    #[test]
    fn malformed_leaf_dropped() {
        let m: Map<String, Value> = json!({"step_visits": {"st-a": 2, "st-b": "no", "st-c": [1]}})
            .as_object()
            .unwrap()
            .clone();
        let sv = get_step_visits(&m);
        assert_eq!(sv.get("st-a"), Some(&2));
        assert!(!sv.contains_key("st-b"));
        assert!(!sv.contains_key("st-c"));
    }
}
